coordinator/src/main/java/com/tencent/rss/coordinator/PartitionBalanceAssignmentStrategy.java

```diff
@@ -0,0 +1,145 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package com.tencent.rss.coordinator;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.tencent.rss.common.PartitionRange;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+public class PartitionBalanceAssignmentStrategy implements AssignmentStrategy {
+
+  private ClusterManager clusterManager;
+  private Map<ServerNode, PartitionInfo> serverToPartitions = Maps.newConcurrentMap();
+
+  public PartitionBalanceAssignmentStrategy(ClusterManager clusterManager) {
+    this.clusterManager = clusterManager;
+  }
+
+  @Override
+  public PartitionRangeAssignment assign(
+      int totalPartitionNum,
+      int partitionNumPerRange,
+      int replica,
+      Set<String> requiredTags) {
+
+    if (partitionNumPerRange != 1) {
+      throw new RuntimeException("PartitionNumPerRange must be one");
+    }
+
+    SortedMap<PartitionRange, List<ServerNode>> assignments = new TreeMap<>();
+    synchronized (this) {
+        List<ServerNode> nodes = clusterManager.getServerList(requiredTags);
+        Map<ServerNode, PartitionInfo> newPartitionInfos = Maps.newConcurrentMap();
+        for (ServerNode node : nodes) {
+          PartitionInfo partitionInfo;
+          if (serverToPartitions.containsKey(node)) {
+            partitionInfo = serverToPartitions.get(node);
+            if (partitionInfo.getTimestamp() < node.getTimestamp()) {
+              partitionInfo.resetPartitionNum();
+              partitionInfo.setTimestamp(node.getTimestamp());
+            }
+          } else {
+            partitionInfo = new PartitionInfo();
+          }
+          newPartitionInfos.putIfAbsent(node, partitionInfo);
+        }
+        serverToPartitions = newPartitionInfos;
+        int assignPartitions = totalPartitionNum * replica / clusterManager.getShuffleNodesMax();
+        for (ServerNode node : nodes) {
+          int p = serverToPartitions.get(node).getPartitionNum();
+        }
+        nodes.sort(new Comparator<ServerNode>() {
+          @Override
+          public int compare(ServerNode o1, ServerNode o2) {
+            PartitionInfo partitionInfo1 = serverToPartitions.get(o1);
+            PartitionInfo partitionInfo2 = serverToPartitions.get(o2);
+            double v1 = o1.getAvailableMemory() * 1.0 / (partitionInfo1.getPartitionNum() + assignPartitions);
+            double v2 = o2.getAvailableMemory() * 1.0 / (partitionInfo2.getPartitionNum() + assignPartitions);
+            return -Double.compare(v1, v2);
+          }
+        });
+        List<ServerNode> assignedNodes = nodes.subList(0, clusterManager.getShuffleNodesMax());
+        int idx = 0;
+        for (int partition = 0; partition < totalPartitionNum; partition++) {
```

Comment:
how about `for (PartitionRange range : ranges)` to avoid create `new PartitionRange(partition, partition)`

Reply:
PartitionBalanceAssignmentStrategy don't support range partition

Reply:
> PartitionBalanceAssignmentStrategy don't support range partition

`partitionNumPerRange != 1` is already checked

Reply:
> how about `for (PartitionRange range : ranges)` to avoid create `new PartitionRange(partition, partition)`

Avoid creating `new PartitionRange(partition, partition)`, But you must create ranges. It's not different.

Reply:
The pr is not target to support range partition, but it should be supported if range partition is still an option. Reuse `BasicAssignmentStrategy.generateRanges()` to generate ranges is for such improvement.